### PR TITLE
clock_control: nrf5: Workaround errata 201

### PR DESCRIPTION
**Pull request by cvinayak**
Add workaround for errata 201, disregard HFCLKSTAT and
EVENT_HFCLKSTARTED after first EVENT_HFCLKSTARTED.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

### CODE REVIEW

**Comment by carlescufi** [comment]
@anangl @nordic-krch can you take a look at this one?

File: drivers/clock_control/nrf5_power_clock.c [hunk 1]
```diff
@@ -17,6 +17,12 @@
 #include <drivers/clock_control/nrf5_clock_control.h>
 #endif
 
+#define HF_STAT_MASK (CLOCK_HFCLKSTAT_SRC_Xtal | CLOCK_HFCLKSTAT_STATE_Msk)
+
```
Comment (nordic-krch) [inline drivers/clock_control/nrf5_power_clock.c]: what about creating local ERRATA_201 define as one may wonder why hacking nrf52.

```
#define ERRATA_201 IS_ENABLED(CONFIG_SOC_SERIES_NRF52X)
```

Than avoid using preprocessor in the code like:
```
if (ERRATA_201) {
...
}
```

**Review by anangl — CHANGES_REQUESTED** [review]
In my opinion this is way too complicated and even hard to read and comprehend. In fact you don't need to check HFCLKSTAT at all since it is sufficient to monitor the HFCLKSTARTED event. Consider using instead changes contained in [this commit](https://github.com/anangl/zephyr/commit/dee46c26835aaa2b473b5110acc74f0164f5c4b1). Similar solution is used successfully in _nrfx_clock.c_.

**Comment by cvinayak** [comment]
@anangl It is important to not monitor the HFCLKSTARTED as the errata 201 states, and only use the first occurrence of the event. Also, the errata is only applicable to nRF52 series, hence the conditional compilations and the use of the check for XTAL source.

**Comment by anangl** [comment]
> @anangl It is important to not monitor the HFCLKSTARTED as the errata 201 states, and only use the first occurrence of the event. Also, the errata is only applicable to nRF52 series, hence the conditional compilations and the use of the check for XTAL source.

By writing "to monitor the HFCLKSTARTED event" I meant to use the first occurrence of this event after issuing the HFCLKSTART task as the indication that the clock has started. This is completely in accordance with the errata (this first event is correct, what you need to do is to disregard all potential subsequent occurrences of the event until the clock is stopped again, but this is not a case in this driver, as the event is not used for executing any callback etc.) and this is actually what is implemented in nRF5 SDK v15.0.0 as the workaround for the anomaly 201.
The advantage of this solution is that you can use it for all SoCs as then you don't need to look at HFLCKSTAT at all.

**Comment by cvinayak** [comment]
@JoeHut sorry missed your comment. if the EVENTS_HFCLKSTARTED does not arrive, then MCU continues to operate using internal RC oscillator wait for the event to be generated (which never happens in your case), this can be faulty xtal or damaged SoC.

**Comment by cvinayak** [comment]
Closing this as there is no activity for few months

**Pull request closed without merging**